File: typeScript/creational-patterns/factoryMethod.ts
interface Notification {
    send(message: string): void;
}

class EmailNotification implements Notification {
    send(message: string): void {
        console.log(`Enviando correo electrónico: ${message}`);
    }
}

class SMSNotification implements Notification {
    send(message: string): void {
        console.log(`Enviando SMS: ${message}`);
    }
}

class PushNotification implements Notification {
    send(message: string): void {
        console.log(`Enviando notificación push: ${message}`);
    }
}

enum NotificationType {
    EMAIL = "email",
    SMS = "sms",
    PUSH = "push",
}

function createNotification(type: NotificationType): Notification {
    switch (type) {
        case NotificationType.EMAIL:
            return new EmailNotification();
        case NotificationType.SMS:
            return new SMSNotification();
        case NotificationType.PUSH:
            return new PushNotification();
        default:
            throw new Error("Tipo de notificación no válido");
    }
}

function main() {
    const emailNotification = createNotification(NotificationType.EMAIL);
    emailNotification.send("¡Hola por correo electrónico!");

    const smsNotification = createNotification(NotificationType.SMS);
    smsNotification.send("¡Hola por SMS!");

    const pushNotification = createNotification(NotificationType.PUSH);
    pushNotification.send("¡Hola por notificación push!");
}

main();

File: rust/creational-patterns/factory_method.rs
use std::fmt;

// Interfaz de notificación
trait Notification {
    fn send(&self, message: &str);
}

// Implementación concreta de notificación por correo electrónico
struct EmailNotification;

impl Notification for EmailNotification {
    fn send(&self, message: &str) {
        println!("Enviando correo electrónico: {}", message);
    }
}

// Implementación concreta de notificación por SMS
struct SMSNotification;

impl Notification for SMSNotification {
    fn send(&self, message: &str) {
        println!("Enviando SMS: {}", message);
    }
}

// Implementación concreta de notificación push
struct PushNotification;

impl Notification for PushNotification {
    fn send(&self, message: &str) {
        println!("Enviando notificación push: {}", message);
    }
}

// Enumeración de tipos de notificación
#[derive(Debug)]
enum NotificationType {
    Email,
    Sms,
    Push,
}

// Error personalizado para tipos de notificación no válidos
#[derive(Debug)]
struct InvalidNotificationTypeError;

impl fmt::Display for InvalidNotificationTypeError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "Tipo de notificación no válido")
    }
}

// Función de fábrica de notificaciones
fn create_notification(notification_type: NotificationType) -> Result<Box<dyn Notification>, InvalidNotificationTypeError> {
    match notification_type {
        NotificationType::Email => Ok(Box::new(EmailNotification)),
        NotificationType::Sms => Ok(Box::new(SMSNotification)),
        NotificationType::Push => Ok(Box::new(PushNotification)),
    }
}

fn main() {
    // Crear una notificación por correo electrónico
    let email_notification = create_notification(NotificationType::Email).unwrap();
    email_notification.send("¡Hola por correo electrónico!");

    // Crear una notificación por SMS
    let sms_notification = create_notification(NotificationType::Sms).unwrap();
    sms_notification.send("¡Hola por SMS!");

    // Crear una notificación push
    let push_notification = create_notification(NotificationType::Push).unwrap();
    push_notification.send("¡Hola por notificación push!");
}
